5548 Provide admin link on incomplete address card reader connection error

joshheald opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: #5548
<!-- Id number of the GitHub issue this PR addresses. -->
Merge #5547 before this

### Description
This changes the behaviour of the primary button on the incomplete address error alert, so that it now opens the `adminUrl` received from the API. 

[Like Android](https://github.com/woocommerce/woocommerce-ios/issues/5088#issuecomment-943220808), this will open the admin pages within an authenticated in-app web view if it's a Wordpress.com store, and send the user to Safari if it's a self-hosted site.
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->

### Questions
1. Getting the site info. I'm not sure where the best place to do this is, I've opted for one class up from the view model for now but there's no particular logic to that decision. Where do you think would be best to make that service locator call? 
2. Is there a better way of finding out whether it's a WP.com site?
3. Have you come across the runtime error `Accessing State's value outside of being installed on a View. This will result in a constant Binding of the initial value and will not update.`? This happens if I try to use the `isPresented` property of `AuthenticatedWebview` outside SwiftUI, and I think it's unavoidable in this situation. If you know otherwise, it might be nicer to be able to use that to manage presentation rather than calling `viewController.dismiss()` in the closure.
4. Handling the return from Safari for the self-hosted sites isn't attempted yet, so any comments on the options in the TODO there would be appreciated.

### Testing instructions
<!-- Step by step testing instructions. When necessary break out individual scenarios that need testing, consider including a checklist for the reviewer to go through. -->
#### Wordpress.com/Atomic stores
1. With a WP.com or Atomic store set up for In-Person Payments, go to `WP-Admin > WooCommerce > Settings` and delete the contents of `Address line 1`
2. In the app, go to `⚙️ > In-Person Payments > Manage Card Reader`
3. Search for and connect to a reader (or simulated reader)
4. Observe that the error message is shown telling the merchant to update their address.
5. Tap `Enter Address` to open a webview at the admin pages, you should be logged in already.
6. Update the address in WP-Admin to include `Address line 1` again
7. Tap `Done`, and connect to the reader again.

#### Self hosted stores
1. With a self-hosted store set up for In-Person Payments, go to `WP-Admin > WooCommerce > Settings` and delete the contents of `Address line 1`
2. In the app, go to `⚙️ > In-Person Payments > Manage Card Reader`
3. Search for and connect to a reader (or simulated reader)
4. Observe that the error message is shown telling the merchant to update their address.
5. Tap `Enter Address` to go to Safari at the admin pages, login if necessary.
6. Update the address in WP-Admin to include `Address line 1` again
7. Return to the app, cancel on the alert, and connect to the reader again

### Screenshots
<!-- Include before and after images or gifs when appropriate. -->
#### Wordpress.com/Atomic store


https://user-images.githubusercontent.com/2472348/144076136-013fe75c-3a8f-4019-82f0-518faa01a0c9.mp4




#### Self hosted stores



https://user-images.githubusercontent.com/2472348/144076191-323cb1bb-566c-4335-9ed7-0527b1cb49b5.mp4


---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

joshheald committed:
5548 Rename action on address error to reflect use

joshheald committed:
5548 Pass adminUrl to incomplete address error UI

joshheald committed:
5548 Add admin button to update address on error

This commit does not deal with webview authentication or refreshing after the webview is closed.

joshheald committed:
5548 Open WPCom admin in-app, otherwise use Safari

ctarda commented on WooCommerce/Classes/ViewModels/CardPresentPayments/CardPresentModalConnectionFailedUpdateAddress.swift:
```diff
@@ -26,13 +36,52 @@ final class CardPresentModalConnectingFailedUpdateAddress: CardPresentPaymentsMo
+            presentAuthenticatedWebview(url: adminUrl, from: viewController)
+        default:
+            UIApplication.shared.open(adminUrl)
+            // TODO: Replace button with refresh? Dismiss the alert? Do the retry when the user returns to the app?
```
I would suggest keeping it as simple as possible in the first iteration. So replacing the button with "refresh" sounds good enough for now?

ctarda commented:
I think this is moving forward in the right direction. To answer (some) of your questions:

> Getting the site info. I'm not sure where the best place to do this is, I've opted for one class up from the view model for now but there's no particular logic to that decision. Where do you think would be best to make that service locator call?

It does not look particularly out of place to me where it is. That being said, I might explore moving one step up, and pass the site to connectingFailedUpdateAddress as a parameter. 

> Is there a better way of finding out whether it's a WP.com site?

Not that I am aware of.

> Have you come across the runtime error Accessing State's value outside of being installed on a View. This will result in a constant Binding of the initial value and will not update.? This happens if I try to use the isPresented property of AuthenticatedWebview outside SwiftUI, and I think it's unavoidable in this situation. If you know otherwise, it might be nicer to be able to use that to manage presentation rather than calling viewController.dismiss() in the closure.

I don't know, sorry.

> Handling the return from Safari for the self-hosted sites isn't attempted yet, so any comments on the options in the TODO there would be appreciated.

Left a comment there. Reducing scope and iterating if needed always sounds like a good option 😄

joshheald committed:
5548 Open adminUrl in Safari for selfhosted stores

The button will be changed to `Refresh After Updating` when they navigate away to Safari.

joshheald committed:
Pass Site from CardReaderConnectionController

allendav commented on WooCommerce/Classes/ViewModels/CardPresentPayments/CardPresentModalConnectionFailedUpdateAddress.swift:
```diff
@@ -26,13 +37,56 @@ final class CardPresentModalConnectingFailedUpdateAddress: CardPresentPaymentsMo
+              let viewController = viewController else {
+            return retrySearchAction()
+        }
+        switch site?.isWordPressStore {
```
Suggestion: Instead of making this aware of Sites and WordPress, could we instead pass a presentation flag?

allendav commented on WooCommerce/Classes/ViewModels/CardPresentPayments/CardPresentModalConnectionFailedUpdateAddress.swift:
```diff
@@ -26,13 +37,56 @@ final class CardPresentModalConnectingFailedUpdateAddress: CardPresentPaymentsMo
+        }
+    }
+
+    private func presentAuthenticatedWebview(url adminUrl: URL, from viewController: UIViewController) {
```
Suggestion: Should the modal be doing this (and aware of AuthenticatedWebView), or could we keep the modal more generic by having the caller deal with this?

allendav reviewed (approved):
Tests fine. In addition to the suggestions I posted in the review, I wonder if we should just check the store address as part of our other many checks (country, account, etc.) instead of getting all the way to connecting to a reader just to find out the setup isn't correct - but that can go in a new issue.

![IMG_0265](https://user-images.githubusercontent.com/1595739/144280256-48b6d78c-347b-41ae-8c58-1f08e7b838e9.jpg)


:shipit:

joshheald commented:
> Tests fine. In addition to the suggestions I posted in the review, I wonder if we should just check the store address as part of our other many checks (country, account, etc.) instead of getting all the way to connecting to a reader just to find out the setup isn't correct - but that can go in a new issue.
> 
> ![IMG_0265](https://user-images.githubusercontent.com/1595739/144280256-48b6d78c-347b-41ae-8c58-1f08e7b838e9.jpg)
> 
> :shipit:

Raised #5591 for this, and #5590 for PR changes
